### PR TITLE
Implements python-cryptography module

### PR DESCRIPTION
**Pull request by lukehinds**
This change removes pycryptodome and replaces it with the
cryptography library.

This is largely to do with the cryptography being a wrapper of
openssl, which is under greater scrutiny for security bugs
than pycryptodome.

cryptography is also widely available as an OS package with all
Linux distributions.

Resolves: #127

### CODE REVIEW

**Comment by lukehinds** [comment]
gah! tpm 1.2

**Comment by lukehinds** [comment]
@nabilschear , hitting the following issue when using python cryptography with tpm1 :

```
    def __get_mod_from_pem(self, pem):
        pubkey = crypto.rsa_import_pubkey(pem)
        return base64.b64encode(bytearray.fromhex('{:0192x}'.format(pubkey.n)))
```

```
Traceback (most recent call last):

  File "/usr/local/bin/keylime_agent", line 11, in <module>

    load_entry_point('keylime==1.2', 'console_scripts', 'keylime_agent')()

  File "/usr/local/lib/python3.7/site-packages/keylime-1.2-py3.7.egg/keylime/cloud_agent.py", line 492, in main

    (ek,ekcert,aik,ek_tpm,aik_name) = tpm.tpm_init(self_activate=False,config_pw=config.get('cloud_agent','tpm_ownerpassword')) # this tells initialize not to self activate the AIK

  File "/usr/local/lib/python3.7/site-packages/keylime-1.2-py3.7.egg/keylime/tpm1.py", line 579, in tpm_init

    self.__create_aik(self_activate)

  File "/usr/local/lib/python3.7/site-packages/keylime-1.2-py3.7.egg/keylime/tpm1.py", line 336, in __create_aik

    mod = self.__get_mod_from_pem(pem)

  File "/usr/local/lib/python3.7/site-packages/keylime-1.2-py3.7.egg/keylime/tpm1.py", line 565, in __get_mod_from_pem

    return base64.b64encode(bytearray.fromhex('{:0192x}'.format(pubkey.n)))

AttributeError: '_RSAPublicKey' object has no attribute 'n'
```
Do you recall what attribute `n` constitutes? I have been looking through the pycryptodome docs, but not worked it out

File: keylime/crypto.py [hunk 1]
```diff
@@ -19,58 +19,119 @@
...
+from cryptography import exceptions
+import cryptography.hazmat.primitives.asymmetric
+from cryptography.hazmat.primitives.ciphers import ( Cipher, algorithms, modes )
+from cryptography.hazmat.primitives import (hashes,serialization )
+from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
+from cryptography.hazmat.primitives.asymmetric import padding
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives.asymmetric.rsa import generate_private_key
+
+aes_block_size = 16
```
Comment (jetwhiz) [inline keylime/crypto.py]: Can this be set to `algorithms.AES.block_size` instead of hard-coded to '16'?

Comment (lukehinds) [reply keylime/crypto.py]: No objections from me.

**Comment by jetwhiz** [comment]
> @nabilschear , hitting the following issue when using python cryptography with tpm1 :
> 
> ```
>     def __get_mod_from_pem(self, pem):
>         pubkey = crypto.rsa_import_pubkey(pem)
>         return base64.b64encode(bytearray.fromhex('{:0192x}'.format(pubkey.n)))
> ```
> 
> ```
> Traceback (most recent call last):
> 
>   File "/usr/local/bin/keylime_agent", line 11, in <module>
> 
>     load_entry_point('keylime==1.2', 'console_scripts', 'keylime_agent')()
> 
>   File "/usr/local/lib/python3.7/site-packages/keylime-1.2-py3.7.egg/keylime/cloud_agent.py", line 492, in main
> 
>     (ek,ekcert,aik,ek_tpm,aik_name) = tpm.tpm_init(self_activate=False,config_pw=config.get('cloud_agent','tpm_ownerpassword')) # this tells initialize not to self activate the AIK
> 
>   File "/usr/local/lib/python3.7/site-packages/keylime-1.2-py3.7.egg/keylime/tpm1.py", line 579, in tpm_init
> 
>     self.__create_aik(self_activate)
> 
>   File "/usr/local/lib/python3.7/site-packages/keylime-1.2-py3.7.egg/keylime/tpm1.py", line 336, in __create_aik
> 
>     mod = self.__get_mod_from_pem(pem)
> 
>   File "/usr/local/lib/python3.7/site-packages/keylime-1.2-py3.7.egg/keylime/tpm1.py", line 565, in __get_mod_from_pem
> 
>     return base64.b64encode(bytearray.fromhex('{:0192x}'.format(pubkey.n)))
> 
> AttributeError: '_RSAPublicKey' object has no attribute 'n'
> ```
> 
> Do you recall what attribute `n` constitutes? I have been looking through the pycryptodome docs, but not worked it out

That's the public modulus for the RSA key: https://cryptography.io/en/latest/hazmat/primitives/asymmetric/rsa/#numbers

This part of the code should be handled by M2Crypto, so I'm not sure why this has broken ...

**Comment by lukehinds** [comment]
Seems like its definitely pycryptodome that was handling it:

https://github.com/keylime/keylime/blob/be47f803fee52a61e4497ffec28890a896c75f11/keylime/tpm1.py#L562-L564

https://github.com/keylime/keylime/blob/be47f803fee52a61e4497ffec28890a896c75f11/keylime/crypto.py#L28-L35

**Comment by jetwhiz** [comment]
Ah you're right, I was looking farther down in the code.  

Since `rsa_import_pubkey` now returns an `RSAPublicKey` object we'll have to call `public_numbers()` on it and then access `n`.

**Comment by lukehinds** [comment]
I spat out `public_numbers()` to see what it holds:

```
public_numbers(): <RSAPublicNumbers(e=65537, n=19449055647301933106209385199898322798596009317083670649332474593254941756024263942165797270871626184814887115268268083948650723177702903989027567503024387863772707514530893513235946039594183784855342806071839870553270572010820057853180025435802435571012130917996491213861183862833879844689525354671859527030618036355470858556310163012796174176700134083102683759154120000675097115447870309630980546954197767703470679703647546123521248249332259356075481148094893011300462997635302524989730772842062086322649621134673051065680394180635769215001961802075275738583369879757452505114915307021657520398134498451611143671111)>
```

Is this something we can use?

**Comment by jetwhiz** [comment]
You want to access `n` there.  Does something like this work? 

````python
 def __get_mod_from_pem(self, pem): 
     pubkey_nums = crypto.rsa_import_pubkey(pem).public_numbers()
     return base64.b64encode(bytearray.fromhex('{:0192x}'.format(pubkey_nums.n))) 
````

**Comment by lukehinds** [comment]
> You want to access `n` there. Does something like this work?
> 
> ```python
>  def __get_mod_from_pem(self, pem): 
>      pubkey_nums = crypto.rsa_import_pubkey(pem).public_numbers()
>      return base64.b64encode(bytearray.fromhex('{:0192x}'.format(pubkey_nums.n))) 
> ```

Brilliant, thanks Charlie, that works now

**Comment by lukehinds** [comment]
@jetwhiz 

```
iv = bytes(ciphertext[:aes_block_size])
tag = bytes(ciphertext[-aes_block_size:])
```

https://travis-ci.org/keylime/keylime/builds/589774727#L1000

In gitter you commented:

> actually, iv might need to be a bytearray instead of bytes, but tag has to be bytes
> it couldn't hurt to explicitly cast the generated AES keys as bytearray also: https://github.com/keylime/keylime/blob/81bdc5bd640f1300dac38f84bfebddfda0b06c73/keylime/crypto.py#L134

do you mean something like this:

```
def generate_random_key(size=32):
    """ Generate random key using urandom wrapper  """
    return os.urandom(bytes(size))
```

**Comment by jetwhiz** [comment]
@lukehinds 

I was thinking something along these lines, though I don't see how this should change anything: 
````python 

def decrypt(ciphertext, key):
    """ Decrypt object """
    ciphertext = base64.b64decode(ciphertext)
    iv = bytearray(ciphertext[:aes_block_size])
    tag = bytes(ciphertext[-aes_block_size:])
    ciphertext = bytes(ciphertext[aes_block_size:-aes_block_size])

    decryptor = Cipher(algorithms.AES(key),
                       modes.GCM(iv, tag),
                       backend=default_backend()).decryptor()
````

and 

````python 
def generate_random_key(size=32):
    """ Generate random key using urandom wrapper  """
    return bytearray(os.urandom(size))
````

Their commit that added bytes-like support (https://github.com/pyca/cryptography/commit/3c68250ad9bfb275c760fcce4c72c78c99b57c34#diff-76c3f2faa5efeba3837b92043bc4524eR459) seems to really want everything to be explicitly cast to bytearray, though they should already already be bytes-like (so I don't see how this should change anything).

**Comment by lukehinds** [comment]
@jetwhiz , last recommended change has broke the build: 

https://travis-ci.org/keylime/keylime/builds/590019407?utm_source=github_status&utm_medium=notification

I won't role back yet, unless you want to make a tweak.

**Comment by jetwhiz** [comment]
Ah, the problem might be that the CI is using cryptography v2.1.4 for some reason, which is really old: 

````
Using /usr/lib64/python3.7/site-packages
Searching for cryptography==2.1.4
Best match: cryptography 2.1.4
Adding cryptography 2.1.4 to easy-install.pth file
````

Fedora 30 has v2.6.1.  Even Debian Buster is on v2.6.1 (https://packages.debian.org/buster/python3-cryptography).  Is there a reason the CI is using such an old package?

File: test/test-requirements.txt [hunk 2]
```diff
@@ -1,6 +1,6 @@
 coverage==4.5.2
 green==2.13.0
-pycryptodomex==3.7.2
+cryptography==2.1.4
```
Comment (jetwhiz) [inline test/test-requirements.txt]: Why is this v2.1.4?

Comment (lukehinds) [reply test/test-requirements.txt]: Well spotted, I have no idea why I used that version.

**Comment by lukehinds** [comment]
Excellent, that solved it. 

I think we can merge now @jetwhiz ?

**Comment by nabilschear** [comment]
does this solve that weird problem that we couldn't track down @lukehinds ?

**Comment by lukehinds** [comment]
> does this solve that weird problem that we couldn't track down @lukehinds ?

Yup.

**Comment by jetwhiz** [comment]
I think it looks good now!

**Pull request merged**